SalePrice - the property's sale price in dollars. target variable to predict.

MSSubClass: The building class

MSZoning: The general zoning classification

LotFrontage: Linear feet of street connected to property

LotArea: Lot size in square feet

Street: Type of road access

Alley: Type of alley access

LotShape: General shape of property

LandContour: Flatness of the property

Utilities: Type of utilities available

LotConfig: Lot configuration

LandSlope: Slope of property

Neighborhood: Physical locations within Ames city limits

Condition1: Proximity to main road or railroad

Condition2: Proximity to main road or railroad (if a second is present)

BldgType: Type of dwelling

HouseStyle: Style of dwelling

OverallQual: Overall material and finish quality

OverallCond: Overall condition rating

YearBuilt: Original construction date

YearRemodAdd: Remodel date

RoofStyle: Type of roof

RoofMatl: Roof material

Exterior1st: Exterior covering on house

Exterior2nd: Exterior covering on house (if more than one material)

MasVnrType: Masonry veneer type

MasVnrArea: Masonry veneer area in square feet

ExterQual: Exterior material quality

ExterCond: Present condition of the material on the exterior

Foundation: Type of foundation

BsmtQual: Height of the basement

BsmtCond: General condition of the basement

BsmtExposure: Walkout or garden level basement walls

BsmtFinType1: Quality of basement finished area

BsmtFinSF1: Type 1 finished square feet

BsmtFinType2: Quality of second finished area (if present)

BsmtFinSF2: Type 2 finished square feet

BsmtUnfSF: Unfinished square feet of basement area

TotalBsmtSF: Total square feet of basement area

Heating: Type of heating

HeatingQC: Heating quality and condition

CentralAir: Central air conditioning

Electrical: Electrical system

1stFlrSF: First Floor square feet

2ndFlrSF: Second floor square feet

LowQualFinSF: Low quality finished square feet (all floors)

GrLivArea: Above grade (ground) living area square feet

BsmtFullBath: Basement full bathrooms

BsmtHalfBath: Basement half bathrooms

FullBath: Full bathrooms above grade

HalfBath: Half baths above grade

Bedroom: Number of bedrooms above basement level

Kitchen: Number of kitchens

KitchenQual: Kitchen quality

TotRmsAbvGrd: Total rooms above grade (does not include bathrooms)

Functional: Home functionality rating

Fireplaces: Number of fireplaces

FireplaceQu: Fireplace quality

GarageType: Garage location

GarageYrBlt: Year garage was built

GarageFinish: Interior finish of the garage

GarageCars: Size of garage in car capacity

GarageArea: Size of garage in square feet

GarageQual: Garage quality

GarageCond: Garage condition

PavedDrive: Paved driveway

WoodDeckSF: Wood deck area in square feet

OpenPorchSF: Open porch area in square feet

EnclosedPorch: Enclosed porch area in square feet

3SsnPorch: Three season porch area in square feet

ScreenPorch: Screen porch area in square feet

PoolArea: Pool area in square feet

PoolQC: Pool quality

Fence: Fence quality

MiscFeature: Miscellaneous feature not covered in other categories

MiscVal: $Value of miscellaneous feature

MoSold: Month Sold

YrSold: Year Sold

SaleType: Type of sale

SaleCondition: Condition of sale


In [39]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [40]:
data = pd.read_csv('Training.csv')

In [41]:
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             999 non-null    int64  
 1   MSSubClass     999 non-null    int64  
 2   MSZoning       999 non-null    object 
 3   LotFrontage    826 non-null    float64
 4   LotArea        999 non-null    int64  
 5   Street         999 non-null    object 
 6   Alley          65 non-null     object 
 7   LotShape       999 non-null    object 
 8   LandContour    999 non-null    object 
 9   Utilities      999 non-null    object 
 10  LotConfig      999 non-null    object 
 11  LandSlope      999 non-null    object 
 12  Neighborhood   999 non-null    object 
 13  Condition1     999 non-null    object 
 14  Condition2     999 non-null    object 
 15  BldgType       999 non-null    object 
 16  HouseStyle     999 non-null    object 
 17  OverallQual    999 non-null    int64  
 18  OverallCon

In [43]:
SalePrice = data["SalePrice"]
data.drop('SalePrice', axis=1, inplace=True)

In [44]:
summary = data.describe(include='object').T.sort_values(by=['count'])
summary

count unique      top freq
PoolQC            2      2       Fa    1
MiscFeature      43      3     Shed   40
Alley            65      2     Grvl   35
Fence           194      4    MnPrv  109
FireplaceQu     522      5       Gd  249
GarageCond      943      5       TA  909
GarageQual      943      5       TA  900
GarageFinish    943      3      Unf  416
GarageType      943      6   Attchd  598
BsmtExposure    974      4       No  653
BsmtFinType2    974      6      Unf  857
BsmtFinType1    975      6      Unf  290
BsmtQual        975      4       TA  432
BsmtCond        975      4       TA  892
MasVnrType      993      4     None  584
Electrical      999      5    SBrkr  909
CentralAir      999      2        Y  931
HeatingQC       999      5       Ex  502
Heating         999      4     GasA  981
Functional      999      7      Typ  931
PavedDrive      999      3        Y  917
KitchenQual     999      4       TA  506
MSZoning        999      5       RL  780
Foundation      999      6    PConc  446
Street          999      2     Pave  995
LotShape        999      4      Reg  622
LandContour     999      4      Lvl  898
Utilities       999      2   AllPub  998
LotConfig       999      5   Inside  715
LandSlope       999      3      Gtl  944
Neighborhood    999     25    NAmes  152
Condition1      999      9     Norm  861
Condition2      999      6     Norm  987
BldgType        999      5     1Fam  836
HouseStyle      999      8   1Story  496
RoofStyle       999      5    Gable  785
RoofMatl        999      6  CompShg  980
Exterior1st     999     11  VinylSd  352
Exterior2nd     999     15  VinylSd  346
ExterQual       999      4       TA  618
ExterCond       999      5       TA  871
SaleType        999      9       WD  859
SaleCondition   999      6   Normal  812

The first 15 inputs/variables hold NAn within, however since this is housing prices majority of these data points could also just mean that the specific location does not support that feature. Like pools for example, are not located in every housing unit.

Same with Garage, each with a count of 943, the remaining few don't have garages therefore they all include an NA.

In [45]:
#data['PoolQC'].unique()
#data['PoolQC'].fillna('NA',inplace=True)

In [46]:
NaN_inputs_categories = []
for i in range(len(summary)):
    if summary.iloc[i][0] < 999:
        NaN_inputs_categories.append(summary.iloc[i].name)
NaN_inputs_categories

['PoolQC',
 'MiscFeature',
 'Alley',
 'Fence',
 'FireplaceQu',
 'GarageCond',
 'GarageQual',
 'GarageFinish',
 'GarageType',
 'BsmtExposure',
 'BsmtFinType2',
 'BsmtFinType1',
 'BsmtQual',
 'BsmtCond',
 'MasVnrType']

In [47]:
for i in NaN_inputs:
    data[i].fillna('NA', inplace = True)

In [48]:
data.describe(include=[np.number]).T.sort_values(by=['count']).head()

count         mean         std     min     25%     50%     75%  \
LotFrontage   826.0    69.962470   23.215821    21.0    60.0    70.0    80.0   
GarageYrBlt   943.0  1978.652174   24.789430  1900.0  1961.0  1980.0  2002.0   
MasVnrArea    993.0   109.311178  188.961644     0.0     0.0     0.0   175.0   
Id            999.0   500.000000  288.530761     1.0   250.5   500.0   749.5   
BedroomAbvGr  999.0     2.855856    0.813983     0.0     2.0     3.0     3.0   

                 max  
LotFrontage    313.0  
GarageYrBlt   2010.0  
MasVnrArea    1600.0  
Id             999.0  
BedroomAbvGr     8.0

In [49]:
data['LotFrontage'].fillna('NA',inplace=True)
data['GarageYrBlt'].fillna('NA',inplace=True)
data['MasVnrArea'].fillna('NA',inplace=True)